### DISCUSSION TITLE
Recommendations on development environment

### DISCUSSION THREAD

**Discussion by mattbaylor** [postalserver/postal#3053] [Development discussions]
I have never written in rails, but I have in a number of other languages. I have an idea about how to tackle https://github.com/orgs/postalserver/discussions/1442, but I need a little guidance on how to setup a dev environment. I was planning on using a linux box, but I have a fair amount of flexibility on what I set up to use. Thanks in advance for the guidance!

**Comment by willpower232** [comment]
Given a docker image is already provided, and managing ruby versions and tooling can be a pain, you could start your own docker-compose.yml and mount the code folder into /opt/postal/app I think.

If you reference https://github.com/postalserver/postal/blob/main/doc/config/environment-variables.md (and maybe https://github.com/postalserver/postal/blob/main/lib/postal/config_schema.rb) then you can configure your dockerised Postal using environment variables within docker-compose.yml.

You will have to provide a MySQL database with root access as well but that is only as annoying as an extra section in docker-compose.yml.

You could start from the install template but you wouldn't need things like host mode networking or the restart unless stopped part
https://github.com/postalserver/install/blob/main/templates/docker-compose.v3.yml

I recently created a mailpit-like solution with this but as it doesn't have persistent storage for the database or the signing key, it could be annoying to develop with. It also isn't quick to start as it downloads maria each time, I probably should make a combined docker image :sweat_smile: I can send queued emails if I need to with `docker compose exec postal postal worker`

```
  postal:
    image: ghcr.io/postalserver/postal:3.3.4
    user: "0:0"
    entrypoint:
      - "/bin/sh"
      - -ecx
      - |
          apt update && apt install -y mariadb-server
          service mariadb start
          mkdir /config && openssl genrsa -out /config/signing.key 1024 && chmod 644 /config/signing.key
          postal initialize
          rails runner "User.create(first_name: 'Admin', last_name: 'User', email_address: 'admin@example.com', password: 'password', admin: true)"
          rails runner "Organization.create(name: 'Test', owner: User.first)"
          rails runner "Organization.first.servers.create(name: 'Test', mode: 'Live')"
          rails runner "Server.first.credentials.create(name: 'Test', type: 'API')"
          rails runner "Server.first.domains.create(name: 'example.com', verification_method: 'DNS', verified_at: Time.now)"
          postal web-server
    ports:
      - "8025:80"
    environment:
      - MAIN_DB_USERNAME=root
      - MESSAGE_DB_USERNAME=root
      - POSTAL_WEB_HOSTNAME=postal
      - POSTAL_WEB_PROTOCOL=http
      - RAILS_ENVIRONMENT=development
      - RAILS_SECRET_KEY=<redacted>
      - WEB_SERVER_DEFAULT_BIND_ADDRESS=0.0.0.0
      - WEB_SERVER_DEFAULT_PORT=80
```
